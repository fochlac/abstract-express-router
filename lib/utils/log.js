/**
 * factory function for the logger
 *
 * loglevels:
 * 0 - silent, 1 - error, 2 - warning, 3 - info, 4 - debug
 *
 * @param {Object} config
 * @returns {(level: number, ...messages:string) => void} configured log function
 */
export const createLogger = ({ logger, logLevel = 3 }) => (level, ...messages) => {
	const message = messages.map(parseMessage).join(' ')
	if (logger) {
		logger({ level, message })
	} else if (level <= logLevel) {
		console.log(`${getDate()}-[abstract-express-router]-${level}-${message}`)
	}
}

function getDate() {
	const now = new Date()

	return `${now.getUTCFullYear()}_${now.getUTCMonth()}_${now.getUTCDate()}-${now.getUTCHours()}:${now.getUTCMinutes()}:${now.getUTCSeconds()}`
}

function parseMessage(item) {
	let output
	if (typeof item === 'string') {
		output = item
	} else {
		try {
			if (Object.prototype.toString.call(item) === '[object Error]') {
				output = '\n' + item.stack.toString()
			} else {
				output = '\n' + JSON.stringify(item, null, 2)
			}
		} catch (err) {
			output = 'error parsing item: ' + item + ' - ' + err
		}
	}
	return output
}
