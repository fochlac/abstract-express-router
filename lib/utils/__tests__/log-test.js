import { spy, stub, useFakeTimers } from 'sinon'

import { createLogger } from '../log'

describe('createLogger', () => {
	useFakeTimers(1554501331074)

	it('should use the provided logger', () => {
		const logger = spy()
		const log = createLogger({ logger })

		log(1, 'test', 'test2', { test: '123' }, new Error('test -split-'))
		log(3, 'test', 'test2')

		expect(logger.callCount).toEqual(2)
		expect(logger.firstCall.args[0].level).toEqual(1)
		expect(logger.firstCall.args[0].message.split('-split-')[0]).toMatchSnapshot()
		expect(logger.secondCall.args).toMatchSnapshot()
	})

	it('should use console.log if no logger is provided', () => {
		stub(console, 'log')
		const log = createLogger({ logLevel: 3 })

		const obj = {test: 'test123'}
		obj.obj = obj
		log(1, 'test', 'test2', { test: '123' }, new Error('test -split-'))
		log(3, 'test', 'test2')
		log(4, 'test', 'test2')
		log(2, 'test', 'test2', obj)

		expect(console.log.callCount).toEqual(3)
		expect(console.log.firstCall.args[0].split('-split-')[0]).toMatchSnapshot()
		expect(console.log.secondCall.args).toMatchSnapshot()
		expect(console.log.thirdCall.args).toMatchSnapshot()
		console.log.restore()
	})
})
