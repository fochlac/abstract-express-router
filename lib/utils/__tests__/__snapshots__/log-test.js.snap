// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLogger should use console.log if no logger is provided 1`] = `
Array [
  "2019_3_5-23:55:31-[abstract-express-router]-1-test test2 
{
  \\"test\\": \\"123\\"
} 
Error: test
    at Object.it (C:\\\\gitHub\\\\json-express-router\\\\lib\\\\utils\\\\__tests__\\\\log-test.js:26:44)
    at Object.asyncJestTest (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:102:37)
    at resolve (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:26:19)
    at promise.then (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:73:41)
    at process._tickCallback (internal/process/next_tick.js:68:7)",
]
`;

exports[`createLogger should use console.log if no logger is provided 2`] = `
Array [
  "2019_3_5-23:55:31-[abstract-express-router]-3-test test2",
]
`;

exports[`createLogger should use console.log if no logger is provided 3`] = `
Array [
  "2019_3_5-23:55:31-[abstract-express-router]-2-test test2 error parsing item: [object Object] - TypeError: Converting circular structure to JSON",
]
`;

exports[`createLogger should use the provided logger 1`] = `
Array [
  Object {
    "level": 1,
    "message": "test test2 
{
  \\"test\\": \\"123\\"
} 
Error: test
    at Object.it (C:\\\\gitHub\\\\json-express-router\\\\lib\\\\utils\\\\__tests__\\\\log-test.js:12:44)
    at Object.asyncJestTest (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:102:37)
    at resolve (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:43:12)
    at new Promise (<anonymous>)
    at mapper (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:26:19)
    at promise.then (C:\\\\gitHub\\\\json-express-router\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:73:41)
    at process._tickCallback (internal/process/next_tick.js:68:7)",
  },
]
`;

exports[`createLogger should use the provided logger 2`] = `
Array [
  Object {
    "level": 3,
    "message": "test test2",
  },
]
`;
